ducc/lib/podman_store.go

```diff
@@ -0,0 +1,499 @@
+//This library creates a Podman image store.
+//Podman image store has the following directory structure:
+
+//	podmanStore
+//	+--	overlay
+//	|	+-- $(layerid)
+//	|	|	+-- diff dir
+//	|	|	+--	link file
+//	|	+-- l
+//	+-- overlay-images
+//	|	+-- $(imageid)
+//	|	|	+-- config file
+//	|	|	+-- manifest file
+//	|	+-- images.json
+//	|	+-- images.lock
+//	+-- overlay-layers
+//	|	+--	layers.json
+//	|	+-- layers.lock
+
+package lib
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+//podman related info stored in image struct.
+type PodmanInfo struct {
+	LayerReaderMap map[string]*ReadAndHash
+	LayerDigestMap map[string]string
+	LayerIdMap     map[string]string
+}
+
+//struct for entries in images.json
+type ImageInfo struct {
+	ID      string    `json:"id,omitempty"`
+	Names   []string  `json:"names,omitempty"`
+	Layer   string    `json:"layer,omitempty"`
+	Created time.Time `json:"created,omitempty"`
+}
+
+//struct for entries in layers.json
+type LayerInfo struct {
+	ID                   string    `json:"id,omitempty"`
+	Parent               string    `json:"parent,omitempty"`
+	Created              time.Time `json:"created,omitempty"`
+	CompressedDiffDigest string    `json:"compressed-diff-digest,omitempty"`
+	CompressedSize       int       `json:"compressed-size,omitempty"`
+	UncompressedDigest   string    `json:"diff-digest,omitempty"`
+	UncompressedSize     int64     `json:"diff-size,omitempty"`
+}
+
+var (
+	//Podman Additional Image Store inside unpacked.cern.ch
+	rootPath = "podmanStore"
+	//rootfsDir contains the exploded rootfs of images.
+	rootfsDir = "overlay"
+	//imageMetadataDir contains the metadata, config and manifest file of images.
+	imageMetadataDir = "overlay-images"
+	//layerMetadataDir contains the metadata of layers
+	layerMetadataDir = "overlay-layers"
+)
+
+//Computes uncompressed digest and size for the given layer
+func (img *Image) ComputeLayerInfo(digest, parent string, layersize int) (layerinfo *LayerInfo, err error) {
+	podmaninfo := img.GetPodmanInfo()
+	r := podmaninfo.LayerReaderMap[digest]
+	diffID := fmt.Sprintf("%x", r.Sum256(nil))
+	size := r.GetSize()
```

Comment:
For this two to be correct, we must be sure that we have consumed all the buffers.

Maybe we should call Read until it doesn't return `0, eof`, what you think?

Reply:
We should also make sure that they somehow coordinates, right?

Reply:
Well, these two methods are called once the layers are completely read and ingested. The call to read function is managed where it's called!
Am I missing something?